test/integration/buffer_accounting_integration_test.cc

```diff
@@ -0,0 +1,226 @@
+#include "test/mocks/http/mocks.h"
+
+#include "fake_upstream.h"
+#include "gtest/gtest.h"
+#include "http_integration.h"
+#include "integration_stream_decoder.h"
+#include "socket_interface_swap.h"
+
+namespace Envoy {
+namespace {
+std::string ipVersionAndBufferAccountingTestParamsToString(
+    const ::testing::TestParamInfo<std::tuple<Network::Address::IpVersion, bool>>& params) {
+  return fmt::format(
+      "{}_{}",
+      TestUtility::ipTestParamsToString(::testing::TestParamInfo<Network::Address::IpVersion>(
+          std::get<0>(params.param), params.index)),
+      std::get<1>(params.param) ? "with_per_stream_buffer_accounting"
+                                : "without_per_stream_buffer_accounting");
+}
+} // namespace
+
+class HttpBufferWatermarksTest
+    : public SocketInterfaceSwap,
+      public testing::TestWithParam<std::tuple<Network::Address::IpVersion, bool>>,
+      public HttpIntegrationTest {
+public:
+  std::vector<IntegrationStreamDecoderPtr>
+  sendRequests(uint32_t num_responses, uint32_t request_body_size, uint32_t response_body_size) {
+    std::vector<IntegrationStreamDecoderPtr> responses;
+
+    Http::TestRequestHeaderMapImpl header_map{
+        {"response_data_blocks", absl::StrCat(1)},
+        {"response_size_bytes", absl::StrCat(response_body_size)},
+        {"no_trailers", "0"}};
+    header_map.copyFrom(default_request_headers_);
+    header_map.setContentLength(request_body_size);
+
+    for (uint32_t idx = 0; idx < num_responses; ++idx) {
+      responses.emplace_back(codec_client_->makeRequestWithBody(header_map, request_body_size));
+    }
+
+    return responses;
+  }
+
+  // TODO(kbaichoo): Parameterize on the client codec type when other protocols
+  // (H1, H3) support buffer accounting.
+  HttpBufferWatermarksTest()
+      : HttpIntegrationTest(Http::CodecClient::Type::HTTP2, std::get<0>(GetParam())) {
+    config_helper_.addRuntimeOverride("envoy.test_only.per_stream_buffer_accounting",
+                                      streamBufferAccounting() ? "true" : "false");
+    setServerBufferFactory(buffer_factory_);
+    setDownstreamProtocol(Http::CodecClient::Type::HTTP2);
+    setUpstreamProtocol(FakeHttpConnection::Type::HTTP2);
+  }
+
+protected:
+  std::shared_ptr<Buffer::TrackedWatermarkBufferFactory> buffer_factory_ =
+      std::make_shared<Buffer::TrackedWatermarkBufferFactory>();
+
+  bool streamBufferAccounting() { return std::get<1>(GetParam()); }
+
+  std::string printAccounts() {
+    std::stringstream stream;
+    auto print_map =
+        [&stream](Buffer::TrackedWatermarkBufferFactory::AccountToBoundBuffersMap& map) {
+          stream << "Printing Account map. Size: " << map.size() << '\n';
+          for (auto& entry : map) {
+            // We can't access the accounts balance in a thread safe way here.
+            stream << "  Account: " << entry.first << '\n';
+            for (auto& buffer : entry.second) {
+              stream << "    Buffer: " << buffer << '\n';
```

Comment:
dumping buffer pointers doesn't provide much info beyond what you'ld get by printing the number of buffers associated with an account.

Is there a real need to track the pointer to buffers associated with the account?